### PR TITLE
Crafting menu mouse support

### PR DESCRIPTION
**Pull request by ZeroInternalReflection**
#### Summary
Interface "Add mouse support to the crafting menu (&)"

#### Purpose of change
The crafting menu is one of many (see #58106) that does not currently have mouse support.

#### Describe the solution
0] Rework how the crafting categories and subcategories are stored/translated to simplify my earlier tab-fitting algorithm (the real reason I started mucking about here).  Also add an optional offset value to draw_tabs() to allow easier mapping if not all tabs are being drawn.

1] Allow selection of crafting category tabs and subtabs with the left mouse button:

https://user-images.githubusercontent.com/89038572/182640958-fafc7cb3-5529-4676-bbb0-f9ff6ee3de0b.mp4

https://user-images.githubusercontent.com/89038572/182640974-f0210322-6e4b-40c2-966a-0d00f1b458fd.mp4

2] Allow scrolling of the recipe list, recipe requirements, and resulting item information using the scrollwheel:

https://user-images.githubusercontent.com/89038572/182641133-081cfad1-4e8f-477b-9cc7-3b74d387b7f1.mp4

https://user-images.githubusercontent.com/89038572/182641147-2680732b-40bc-4c37-95be-ef28a02119b1.mp4

3] Allow selection and crafting of a particular recipe with the left mouse button:

https://user-images.githubusercontent.com/89038572/182641312-b733e7cd-da5f-443f-8582-a56816721807.mp4

4] Ensure mouse-click actions work in curses as well as tiles:

https://user-images.githubusercontent.com/89038572/182641382-05e80500-43c0-4aad-9a12-a7085011a490.mp4

https://user-images.githubusercontent.com/89038572/182641403-3e53aa2a-46ed-477c-91ab-62a3c5071283.mp4

#### Describe alternatives you've considered
1. Not having mouse support?
2. Only having mouse support in tiles, rather than having it work in Curses as well?

There are several UX decisions I've made that could be debated:
1. Clicking on a recipe will mark it as read, scrolling past it with the mousewheel will not.
2. Mousewheel scrolling of the middle and right panes is not as fast as using the keyboard.  I initially tried having it the same speed, and it felt way too fast.  Both could probably be tweaked.
3. I have not added mouseover selection of tabs.  It feels like it would be too easy to mouse over a category, then accidentally mouse over a subcategory you didn't want, then need to go all the way to the left... 
4. I have not added mouseover selection of recipes.  I'm experimenting in a separate PR with having both mousewheel scrolling and mouseover selection enabled on a large list, and it's a bit twitchy.  I'm trying to find a way to only implement the required UX tweaking (e.g. delay on mouseover after scrolling, limiting speed of scrolling from mousing over the bottom of the list, etc.) once, rather than just copying it everywhere.
5. I've also added positional mousewheel scrolling for the header tabs and subtabs as well.  It's a bit much, but without it, there are a couple of category/window size combinations where you can get stuck.  Since you can't select empty subcategories, you might not be able to see the full list of subcategories on a small screen without `Left`/`Right`/mousewheel scrolling

#### Testing
As shown in the videos above, I've tested in several languages at a variety of window sizes.  There's no guarantee that I haven't missed an oddball size that breaks something though.
All tests have been done on Linux, and all Curses tests were done in gnome-terminal.  Testing in other terminals would be appreciated.

#### Additional context
- I'm not sure exactly where the mousewheel action gets lost in curses.  It seems like it should work (at least on gnome-terminal.  Other terminals may not be aware of mouse position).
- Having an individual inclusive_rectangle for each recipe feels wasteful and could probably be simplified by just calculating relevant row information, but it felt responsive enough in testing.  I also still hold out hope of having long recipe names fold to multiple lines, which will complicate everything.

### CODE REVIEW

File: src/crafting_gui.cpp [hunk 1]
```diff
@@ -1338,14 +1395,76 @@ const recipe *select_crafting_recipe( int &batch_size_out, const recipe_id goto_
...
+                    }
+                }
+            }
+        }
+
+        if( action == "SELECT" ) {
+            bool handled = false;
+            if( coord.has_value() ) {
+                point local_coord = coord.value() - point( getbegx( w_head_tabs ), getbegy( w_head_tabs ) );
+                if( !translated_tab_map.empty() ) {
```
Comment (BevapDin) [inline src/crafting_gui.cpp:1436]: You don't have to check this before iterating over the a container. The loop works fine even if the container is empty.

**Comment by Qrox** [comment]
> where the mousewheel action gets lost in curses

On the MSYS2 curses simulator mousewheel action is treated as up/down keys, not sure about other terminals.

File: src/crafting_gui.cpp [hunk 2]
```diff
@@ -1415,7 +1538,7 @@ const recipe *select_crafting_recipe( int &batch_size_out, const recipe_id goto_
             user_moved_line = highlight_unread_recipes;
         } else if( action == "CONFIRM" ) {
             if( available.empty() || !available[line].can_craft ) {
-                popup( _( "You can't do that!  Press [<color_yellow>ESC</color>]!" ) );
+                popup_getkey( _( "You can't do that!  Press [<color_yellow>ESC</color>]!" ) );
```
Comment (Qrox) [inline src/crafting_gui.cpp:1541]: Why is this changed?

Comment (ZeroInternalReflection) [reply src/crafting_gui.cpp:1541]: When I tested it, the original popup would only close with `Escape` or equivalent, while popup_getkey will close with a mouse click.  So, if someone is primarily using the mouse and double-clicks on a recipe that they can't make, the only way to get out of the popup() is to reach over to the keyboard.  It's a minor thing, but it's something that I personally would find annoying.

If the use of popup_getkey stands, the text should probably be updated to 'press any key' or something like that.

Comment (Qrox) [reply src/crafting_gui.cpp:1541]: You could probably use a `query_popup` with `allow_cancel` and a single `QUIT` option, which has been recently changed to support mouse click if I remember correctly.

**Comment by ZeroInternalReflection** [comment]
Per Qrox's suggestion, I've moved to query_popup() for the unable-to-craft warning message.  Removing the superfluous "Press [ESC]" does unfortunately break translation.  Also, getting query_popup mouseclick support in curses is now on my to-do list, but that's a separate PR.

Query_popup_in_crafting_menu

I've also changed the other simple popups in the crafting menu to the same system.  I didn't change them initially since there's no way to get _into_ them without using the keyboard, so I figured needing to use the keyboard to get _out_ of them was less of an annoyance.  However, if I'm changing the look of one of them, I'm changing the look of all of them.

Similar_popup--Broken_translation

I also fixed a bug in mousewheel scrolling where the area for scrolling didn't properly account for the window not being full-width on wide screens.

And, with any luck, I didn't screw up the change clang-tidy was requesting.

**Comment by leemuar** [comment]
Awesome job, thank you!

**Comment by leemuar** [comment]
> Clicking on a recipe will mark it as read, scrolling past it with the mousewheel will not.

This is good, I agree with that. When I was working on mission menu and general UI library I though about a different approach: mouse scroll will not affect a list cursor, it will just scroll the list, not the cursor (just like it happens in modern list UI elements)

> I have not added mouseover selection of tabs/subtabs

I personally support this. Activating elements on hovering is very debatable, and I personally like "highlight on hover, activate on click" approach

**Comment by ZeroInternalReflection** [comment]
> This is good, I agree with that. When I was working on mission menu and general UI library I though about a different approach: mouse scroll will not affect a list cursor, it will just scroll the list, not the cursor (just like it happens in modern list UI elements)

I do like the idea of mouswheel scrolling just the list view but not actually changing selection.  Unfortunately, that means the mousewheel is only relevant for screens like this with a large number of items listed.  Currently, uilist menus use mousewheel scrolling for selection, and changing that would render the mousewheel useless for most ulist menus.
Currently, I'm inclined to leave the behaviour as-is, but if I see a good way to improve mousewheel behaviour (while keeping it consistent) then I'll work on it.
